include/internal/refcount.h

```diff
@@ -36,21 +51,51 @@ static inline int CRYPTO_UP_REF(CRYPTO_REF_COUNT *refcnt, int *ret)
     return 1;
 }
 
-/*
- * Changes to shared structure other than reference counter have to be
- * serialized. And any kind of serialization implies a release fence. This
- * means that by the time reference counter is decremented all other
- * changes are visible on all processors. Hence decrement itself can be
- * relaxed. In case it hits zero, object will be destructed. Since it's
- * last use of the object, destructor programmer might reason that access
- * to mutable members doesn't have to be serialized anymore, which would
- * otherwise imply an acquire fence. Hence conditional acquire fence...
+/**
+ * @brief Atomically decrements the reference count and stores the result.
+ *
+ * This function atomically decreases the reference count of a CRYPTO_REF_COUNT
+ * object. It ensures memory ordering constraints based on the architecture.
+ *
+ * @param refcnt A pointer to the CRYPTO_REF_COUNT object to be decremented.
+ * @param ret A pointer to an integer where the decremented result is stored.
+ *
+ * @return Always returns 1.
+ *
+ * Architecture-specific memory ordering is applied:
+ * - On ARM and AArch64, it uses __ATOMIC_ACQ_REL.
+ * - On other architectures, it uses __ATOMIC_RELAXED and ThreadSanitizer
+ *   acquire/release barriers.
+ *
+ * If the resulting reference count is 0, a memory fence is issued to ensure
+ * proper ordering.
+ *
+ * NOTE: Memory ordering is arch dependent here.  On strongly ordered systems
+ * a relaxed ordering is acceptable, because on such systems (like x86), the
+ * hardware ensures that atomic writes are visible to other cpus immediately.
+ *
+ * However, on weakly ordered systems, and atomic read/write is only atomic with
+ * respect to the local cpu.  To ensure that other cpus see a modified value, we
+ * need to use an atomic aquire/release ordering here, which emits the
+ * appropriate ldxar and stlxr instructions to ensure load/acquire, and
+ * store/release semantics.
+ *
+ * Note further that on x86 platforms, the subsequent atomic fence instruction
+ * isn't interpreted properly by thread sanitizers, and so, with relaxed
+ * ordering, tsan still potentially complains about multithreaded read/write
+ * races.  To avoid incurring the stricter semantic constraints there, we keep
+ * the relaxed ordering (sine the hardware enforces it anyway), and instead
+ * inform tsan that its ok with the __tsan_acquire and __tsan_release notations
  */
 static inline int CRYPTO_DOWN_REF(CRYPTO_REF_COUNT *refcnt, int *ret)
```

Comment:
Sorry but I don't understand what are you trying to do.
There's single correct way to do this
ifdef tsan
fetch_sub(1, acq_rel)
else
if (fetch_sub(1, release)) fence(acquire)
endif

It's simple, correct, and optimal at least for all arm and x86

Reply:
I really don't think thats correct.  Based on my previous comments:

1) We definitely shouldn't be doing a different memory order just for tsan.  If we need acq_rel, then we need it, not just to avoid the tsan errors (which are erroneous since tsan doesn't understand fences).  The only thing that we should do special for tsan is to annotate the fact that we might be using a fence here (i.e. forcibly noting that the memory operation does an acquire and release) because tsan can't otherwise see it.

2) A release ordering followed by an acquire fence makes no sense on weakly memory ordered systems.  The release ordering in that scenario ensures that any prior reads are completed and visible prior to this write being executed, and the acquire fence ensure that any prior writes are completed before any subsequent read is preformed.  We're not doing a subsequent read, so the fence is useless.  The release ordering makes some sense, but doesn't protect against another cpu doing a read of the same memory value at the same time (i.e. the two threads decrementing from 2->1 scenario I proposed)

Reply:
2. That's wrong
1. I don't agree

Reply:
I think our discussion about this code is stuck.
So I don't think you need to answer me.

You can't change my thoughts about this.
I think I understand it's right, and it's too common pattern, and everyone except openssl uses it's that way.


At the same time, I probably wasn't good enough in explaining my thoughts, so I can't change your mind

Reply:
btw even Linux with different (compared to c++) memory model uses same logic:

release fetch_sub + acquire fence 
(ofc code more difficult because it saturated math)
https://github.com/torvalds/linux/blob/master/include/linux/refcount.h#L264
https://github.com/torvalds/linux/blob/master/include/linux/refcount.h#L270

Reply:
I strongly believe that we should **always** specify acquire/release here and forget about the fences.

Furthermore, are we certain that all architectures we support apart from ARM and x86 provide the very strong x86 memory guarantees?  If we must optimise this, we should be checking for x86 and defaulting to the ARM code which will work universally.

Have we measured the performance difference between acquire/release on x86 and the fenced version?  Is there a performance difference in favour of the latter?  Without such a measurement indicating a performance gain, this is premature optimisation.

Reply:
so, in an effort to find common understanding, looking at the kernel refcount code.  It does an atomic_fetch_sub_release, which waits for any prior scheduled store to that address to complete, then does its atomic fetch/sub operation.  After, if the result is non-zero and equal to the test value, it issues an acquire fence, which according to the documentation is actually there to avoid code hoisting of a free operation above the refcount atomic:
```
/**
 * refcount_sub_and_test - subtract from a refcount and test if it is 0
 * @i: amount to subtract from the refcount
 * @r: the refcount
 *
 * Similar to atomic_dec_and_test(), but it will WARN, return false and
 * ultimately leak on underflow and will fail to decrement when saturated
 * at REFCOUNT_SATURATED.
 *
 * Provides release memory ordering, such that prior loads and stores are done
 * before, and provides an acquire ordering on success such that free()
 * must come after.
 *
 * Use of this function is not recommended for the normal reference counting
 * use case in which references are taken and released one at a time.  In these
 * cases, refcount_dec(), or one of its variants, should instead be used to
 * decrement a reference count.
 *
 * Return: true if the resulting refcount is 0, false otherwise
 */
```

If that comment is accurate, and the fence exists only to prevent code sinking/hoisting (which makes more sense than enforcing ordering of subsequent release fences, then I'm struggling to see how this patch doesn't address the same needs.  That is to say an ACQ_REL order will prevent the same code hoisting/sinking, and provides the correct ordering semantics.

Reply:
> After, if the result is non-zero and equal to the test value, it issues an acquire fence, which according to the documentation is actually there to avoid code hoisting of a free operation above the refcount atomic:

For openssl usecase it's same as `if (prev_value - 1 == 0) { fence(acquire); some_openssl_free(...) }`

> If that comment is accurate, and the fence exists only to prevent code sinking/hoisting
Well exists different memory model, maybe in linux memory model it's ok to say code sink/hoist.
But C++ memory model (is C memory model same?) it's different, and you need to talk about synchonize with and happens before relations.

Also I don't think compiler really move loads/stores up/down in case of atomic (well it can technically, but probability is low).

> That is to say an ACQ_REL order will prevent the same code hoisting/sinking, and provides the correct ordering semantics.

Ofc, `fetch_sub(1, acq_rel)` is also completely correct (but in theory a little stronger than necessary), I never said opposite.

> how this patch doesn't address the same needs. 

I think `#else // not arm` branch is wrong.

`*ret = atomic_fetch_sub_explicit(&refcnt->val, 1, __ATOMIC_ACQ_REL) - 1;`

This code from your patch is good enough in my opinion and fix corresponding issue.

Reply:
@nhorman Is any reason why don't you want to put always acq_rel?
x86 will generate completely same instruction, no matter what memory order you will write for rmw-operation